File: src/test/scala/com/scala/scogen/datagenerator/EmployeeDetailsDatasetTest.scala
package com.scala.scogen.datagenerator

import com.scala.scogen.BaseClassTest
import com.scala.scogen.metaData.EmployeeNameList
import org.apache.spark.sql.{Dataset, Row}
import org.junit.Assert

class EmployeeDetailsDatasetTest extends BaseClassTest{

  var employeeDeptDataset:Dataset[Row] = spark.emptyDataFrame
  import spark.implicits._

  def initilize(): Unit =
  {
    if(employeeDeptDataset.isEmpty) {
      val listOfEmployees = EmployeeNameList.setofUniqueNames
      val employeeDs = new EmployeeDetailsDataset().employeeTable(spark, listOfEmployees)

      val eids = employeeDs.rdd.map(x => x.getAs[String]("empid")).collect().toList
      employeeDeptDataset = new EmployeeDeptDataset().empDeptDataset(spark, eids).cache()
    }
  }

  "dept employee count" should "equal 1000" in {
    initilize()
    Assert.assertEquals(1000, employeeDeptDataset.count)
  }

  "number of unique dept" should "be equals 5" in
  {
    initilize()
    Assert.assertEquals(5, employeeDeptDataset.selectExpr("deptid").distinct.count)
  }

  "tow of the depts " should "have 500 employees" in
  {
    initilize()
    val count = employeeDeptDataset
      .selectExpr("deptid").filter($"deptid" === "IT" || $"deptid" === "HR")
      .count
    Assert.assertEquals(500, count)
  }




}

File: src/test/java/com/java/scogen/dataGenerator/EmployeeDetailsDatasetTest.java
package com.java.scogen.dataGenerator;

import com.java.scogen.SparkTestSessionGenerator;
import com.java.scogen.metaData.EmployeeNameList;
import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;
import java.util.Set;

public class EmployeeDetailsDatasetTest extends SparkTestSessionGenerator {

    private Dataset<Row> employeeDeptDataset = null;

    private void initilize()
    {
        if(employeeDeptDataset == null) {
            EmployeeNameList employeeNameList = new EmployeeNameList();
            Set<String> nameSet = employeeNameList.setofUniqueNames();
            EmployeeDetailsDataset employeeDetailsDataset = new EmployeeDetailsDataset(getSpark());
            Dataset<Row> employee = employeeDetailsDataset.EmployeeTable(nameSet);
            List<String> ids = employee.select("empid").toJavaRDD().map(x -> x.<String>getAs("empid"))
                    .collect();
            employeeDeptDataset = new EmployeeDeptDataset(getSpark()).empDeptDataset(ids).cache();
        }

    }
    @Test
    public void deptEmployeeCountTest()
    {
        initilize();
        Assert.assertEquals(1000,employeeDeptDataset.count());
    }

    @Test
    public void deptCountTest()
    {
        initilize();
        Assert.assertEquals(5,employeeDeptDataset.selectExpr("deptid").distinct().count());
    }

    @Test
    public void dept500Test()
    {
        initilize();
        long count = employeeDeptDataset.selectExpr("deptid").filter(new Column("deptid").equalTo("IT")
                .or(new Column("deptid").equalTo("INFRA"))).count();
        Assert.assertEquals(500,count);
    }

}
